src/Compilers/CSharp/Portable/Binder/Binder_TupleOperators.cs

```diff
@@ -0,0 +1,369 @@
+        {
+            ImmutableArray<TypeSymbol> convertedTypes = operators.SelectAsArray((o, r) => r ? o.RightConvertedType : o.LeftConvertedType, isRight);
+            ImmutableArray<Location> elementLocations = elements.SelectAsArray(e => e.Syntax.Location);
+
+            var tuple = TupleTypeSymbol.Create(locationOpt: null, elementTypes: convertedTypes,
+                elementLocations, elementNames: default, compilation,
+                shouldCheckConstraints: true, errorPositions: default, syntax, diagnostics);
+
+            if (!isNullable)
+            {
+                return tuple;
+            }
+
+            NamedTypeSymbol nullableT = GetSpecialType(SpecialType.System_Nullable_T, diagnostics, syntax);
+            return nullableT.Construct(tuple);
+        }
+    }
+
+    /// <summary>
+    /// A tree of binary operators
+    /// </summary>
+    internal abstract class TupleBinaryOperatorInfo
+    {
+        internal abstract bool IsSingle();
+        internal readonly TypeSymbol LeftConvertedType;
+        internal readonly TypeSymbol RightConvertedType;
+#if DEBUG
+        internal abstract TreeDumperNode DumpCore();
+        internal string Dump() => TreeDumper.DumpCompact(DumpCore());
+#endif
+
+        private TupleBinaryOperatorInfo(TypeSymbol leftConvertedType, TypeSymbol rightConvertedType)
+        {
+            LeftConvertedType = leftConvertedType;
+            RightConvertedType = rightConvertedType;
+        }
+
+        internal class Single : TupleBinaryOperatorInfo
+        {
+            internal readonly BinaryOperatorKind Kind;
+            internal readonly MethodSymbol MethodSymbolOpt; // User-defined comparison operator, if applicable
+
+            // To convert the result of comparison to bool
+            internal readonly Conversion BoolConversion;
+            internal readonly UnaryOperatorSignature BoolOperator; // Information for op_true or op_false
+
+            internal Single(TypeSymbol leftConvertedType, TypeSymbol rightConvertedType, BinaryOperatorKind kind,
+                MethodSymbol methodSymbolOpt, Conversion boolConversion, UnaryOperatorSignature boolOperator) : base(leftConvertedType, rightConvertedType)
+            {
+                // If a user-defined comparison operator is present, then the operator kind must be user-defined
+                Debug.Assert(Kind.IsUserDefined() || (object)MethodSymbolOpt == null);
+
+                // If the return type of methodSymbolOpt is not bool, then there must be a boolConversion or boolOperator
+                Debug.Assert(BoolConversion != default || !Kind.IsDynamic() || (Kind.IsUserDefined() && MethodSymbolOpt.ReturnType.SpecialType == SpecialType.System_Boolean));
+                Debug.Assert((object)BoolOperator != null || !Kind.IsDynamic() || (Kind.IsUserDefined() && MethodSymbolOpt.ReturnType.SpecialType == SpecialType.System_Boolean));
```

Comment:
`BoolConversion` and `BoolOperator` are not set yet.

Reply:
Thanks. I ended up removing those assertions, as they were complicated to understand.

---
In reply to: [167979141](https://github.com/dotnet/roslyn/pull/23356#discussion_r167979141) [](ancestors = 167979141)